serverless/log-email-sns/src/index.ts

```diff
@@ -0,0 +1,100 @@
+
+    switch (notificationType) {
+      case "Delivery": 
+        await updateSuccessfulDelivery(message, sequelize)
+        break
+      case "Bounce":
+        await updateBouncedStatus(message, sequelize)
+        break
+      default:
+        throw new Error(`Can't handle messages with this notification type. notificationType = ${notificationType}`)
+    }    
+    
+    return {
+      statusCode: 200,
+      body: 'Ok'
+    }
+
+  } catch(err) {
+    console.error(`Unhandled server error  ${err.name}: ${err.message}`)
+    console.error(`Event: ${JSON.stringify(event)}`)
+
+    return {
+      statusCode: 500
+    }
+  }
+}
+
+/**
+ *  Updates the email_messages table for successful delivery of an email.
+ *  Delivery: Amazon SES successfully delivered the email to the recipient's mail server.
+ *  @param message JSON object that contains the notification details
+ *  @param dbConnection Sequelize connection
+ */
+const updateSuccessfulDelivery = async (message: any, dbConnection: Sequelize ) => {
+  const messageId = message?.mail?.commonHeaders?.messageId
+  const timeStamp = message?.delivery?.timestamp
+
+  await dbConnection.query(
+    `UPDATE email_messages SET received_at=:timeStamp, updated_at = clock_timestamp(), status='SUCCESS' WHERE message_id=:messageId`,
+    {
+      replacements: { timeStamp, messageId }, type: QueryTypes.UPDATE 
+    })
+}
+
+/**
+ *  Updates the error_code in the email_messages table for bounced delivery.
+ *  Hard bounce: The recipient's mail server permanently rejected the email.
+ *  Soft bounce: SES fails to deliver the email after retrying for a period of time.
+ *  @param message JSON object that contains the notification details
+ *  @param dbConnection Sequelize connection
+ */
+const updateBouncedStatus = async (message: any, dbConnection: Sequelize) => {
+  const bounceType = message?.bounce?.bounceType
+  const messageId = message?.mail?.commonHeaders?.messageId
+  const timeStamp = message?.delivery?.timestamp
+  const recipientEmail = message?.mail?.commonHeaders?.to[0]
+
+  let errorCode
+ 
+  if (bounceType === 'Permanent') {
+    errorCode = "Hard bounce, the recipient's mail server permanently rejected the email."
+
+    console.log('Updating blacklist table.')
+    // Add to black list
+    await dbConnection.query(
+      `UPDATE email_blacklist SET recipient=:recipientEmail, updated_at=clock_timestamp() WHERE message_id=:messageId`,
+      {
+        replacements: { recipientEmail, messageId }, type: QueryTypes.UPDATE 
+      })
+  }
+  else {
+    errorCode = "Soft bounce, Amazon SES fails to deliver the email after retrying for a period of time."
+  }
+  
+  console.log('Updating email_messages table.')
+  await dbConnection.query(
+    `UPDATE email_messages SET error_code=:errorCode, received_at=:timeStamp, status='INVALID_RECIPIENT', updated_at = clock_timestamp() WHERE message_id=:messageId`,
```

Comment:
So for soft bounce and permanent bounce i updated the status as `INVALID_RECIPIENT` because I vaguely remember that we will allow resends for `ERROR`. Tbh i am not sure soft bounces should be allowed to resend.